test/integration_test.go

```diff
@@ -873,6 +874,39 @@ func TestTerragruntOutputFromRemoteState(t *testing.T) {
 
 	assert.True(t, (strings.Index(output, "app3 output") < strings.Index(output, "app1 output")) && (strings.Index(output, "app1 output") < strings.Index(output, "app2 output")))
 }
+
+func TestTerragruntOutputFromDependency(t *testing.T) {
+	// you cannot use t.Parallel() with t.Setenv
+
+	fixtureAppPath := filepath.Join(TEST_FIXTURE_OUTPUT_FROM_DEPENDENCY, "app")
+	fixtureDepPath := filepath.Join(TEST_FIXTURE_OUTPUT_FROM_DEPENDENCY, "dependency")
+
+	s3BucketName := fmt.Sprintf("terragrunt-test-bucket-%s", strings.ToLower(uniqueId()))
+	defer deleteS3Bucket(t, TERRAFORM_REMOTE_STATE_S3_REGION, s3BucketName)
+
+	rootTerragruntConfigPath := util.JoinPath(fixtureDepPath, config.DefaultTerragruntConfigPath)
+	copyTerragruntConfigAndFillPlaceholders(t, rootTerragruntConfigPath, rootTerragruntConfigPath, s3BucketName, "not-used", TERRAFORM_REMOTE_STATE_S3_REGION)
+
+	var (
+		stdout bytes.Buffer
+		stderr bytes.Buffer
+	)
+
+	t.Setenv("AWS_CSM_ENABLED", "true")
+
+	err := runTerragruntCommand(t, fmt.Sprintf("terragrunt init --terragrunt-non-interactive --terragrunt-working-dir %s --terragrunt-log-level debug", fixtureDepPath), &stdout, &stderr)
+	assert.NoError(t, err)
+
+	err = runTerragruntCommand(t, fmt.Sprintf("terragrunt apply -auto-approve --terragrunt-non-interactive --terragrunt-working-dir %s --terragrunt-log-level debug", fixtureDepPath), &stdout, &stderr)
+	assert.NoError(t, err)
+
+	err = runTerragruntCommand(t, fmt.Sprintf("terragrunt apply -auto-approve --terragrunt-non-interactive --terragrunt-working-dir %s --terragrunt-log-level debug", fixtureAppPath), &stdout, &stderr)
+	assert.NoError(t, err)
```

Comment:
Terragrunt support auto init and automatic apply of dependencies, I think these invocations can be replaced with:
```
"cd test/fixture-output-from-dependency"
terragrunt run-all apply 
```
It will automatically init and apply first dependency and then app with passing outputs

Examples:
https://github.com/gruntwork-io/terragrunt/blob/master/test/integration_test.go#L3619

Reply:
Yes, I read the documentation and in the first implementation I did exactly as you suggested, but the dependency was not being applied. I just tried to investigate, and look what I figured out:

When I don't use the `--terragrunt-non-interactive` flag, the behavior is as expected: 

```
terragrunt run-all apply --terragrunt-log-level debug                                                                                                                          1 ↵
DEBU[0000] Did not find any locals block: skipping evaluation.  prefix=[/private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app]
DEBU[0000] Setting download directory for module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app to /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app/.terragrunt-cache
DEBU[0000] Did not find any locals block: skipping evaluation.  prefix=[/private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency]
DEBU[0000] Setting download directory for module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency to /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency/.terragrunt-cache
Module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app depends on module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency, which is an external dependency outside of the current working directory. Should Terragrunt run this external dependency? Warning, if you say 'yes', Terragrunt will make changes in /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency as well! (y/n) y
DEBU[0005] Stack at /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app:
  => Module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app (excluded: false, assume applied: false, dependencies: [/private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency])
  => Module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency (excluded: false, assume applied: false, dependencies: [])
INFO[0005] The stack at /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app will be processed in the following order for command apply:
Group 1
- Module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency

Group 2
- Module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app

Are you sure you want to run 'terragrunt apply' in each folder of the stack described above? (y/n)
```

and with `--terragrunt-non-interactive`

```
terragrunt run-all apply --terragrunt-non-interactive --terragrunt-log-level debug                                                                                           130 ↵
DEBU[0000] Did not find any locals block: skipping evaluation.  prefix=[/private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app]
DEBU[0000] Setting download directory for module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app to /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app/.terragrunt-cache
DEBU[0000] Did not find any locals block: skipping evaluation.  prefix=[/private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency]
DEBU[0000] Setting download directory for module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency to /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency/.terragrunt-cache
DEBU[0000] The --non-interactive flag is set. To avoid accidentally affecting external dependencies with a run-all command, will not run this command against module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency, which is a dependency of module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app.
DEBU[0000] Stack at /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app:
  => Module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app (excluded: false, assume applied: false, dependencies: [/private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency])
  => Module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency (excluded: false, assume applied: true, dependencies: [])
INFO[0000] The stack at /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app will be processed in the following order for command apply:
Group 1
- Module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app

DEBU[0000] Are you sure you want to run 'terragrunt apply' in each folder of the stack described above? (y/n)
DEBU[0000] The non-interactive flag is set to true, so assuming 'yes' for all prompts
DEBU[0000] Module /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app must wait for 1 dependencies to finish  prefix=[/private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app]

... omitted

ERRO[0002] 1 error occurred:
        * /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/dependency/terragrunt.hcl is a dependency of /private/work/contract/gruntwork/repos/terragrunt/test/fixture-output-from-dependency/app/terragrunt.hcl but detected no outputs. Either the target module has not been applied yet, or the module has no outputs. If this is expected, set the skip_outputs flag to true on the dependency block.

ERRO[0002] Unable to determine underlying exit code, so Terragrunt will exit with error code 1
```

Looks like a bug, doesn't it?

Reply:
AFAIK, this is expected, since terragrunt is invoked in `fixture-output-from-dependency/app` but dependency is referencing something outside of `app`

To avoid this, invoke terragrunt in `fixture-output-from-dependency` and it will build right dependencies since `app` and `dependency` are child directories

Reply:
Yeah, this helped, thanks! 
Either way, this is a bit of a weird behavior. I think it should behave the same whether or not the `--terragrunt-non-interactive` flag is specified.